src/misc/unicode.cpp

```diff
@@ -32,416 +32,618 @@
-// UTF-8 -> UTF-8 fallback mapping (alias), use before fallback to 7-bit ASCII
+
+// Unicode -> Unicode fallback mapping (alias),
+// use before fallback to 7-bit ASCII
 static config_aliases_t config_aliases = {};
 // Information about code pages which are exact duplicates
 static config_duplicates_t config_duplicates = {};
 
-// UTF-8 -> 7-bit ASCII mapping, use as a last resort mapping
+// Unicode -> 7-bit ASCII mapping, use as a last resort mapping
 static code_page_mapping_t mapping_ascii = {};
 
-// Concrete UTF-8 -> codepage mappings
-static std::map<uint16_t, code_page_mapping_t> mappings_by_codepage = {};
-// Additional UTF-8 -> codepage mappings, to avoid unknown characters
-static std::map<uint16_t, code_page_mapping_t> aliases_by_codepage = {};
-// Reverse mappings, codepage -> UTF-8
-static std::map<uint16_t, code_page_mapping_reverse_t> reverse_by_codepage = {};
+// Unicode 'KD' decomposition rules
+static decomposition_rules_t decomposition_rules = {};
+
+// Concrete Unicode -> codepage mappings
+static std::map<uint16_t, code_page_mapping_t> mappings_normalized_by_codepage = {};
+static std::map<uint16_t, code_page_mapping_t> mappings_decomposed_by_codepage = {};
+// Additional Unicode -> codepage mappings, to avoid unknown characters
+static std::map<uint16_t, code_page_mapping_t> aliases_normalized_by_codepage = {};
+static std::map<uint16_t, code_page_mapping_t> aliases_decomposed_by_codepage = {};
+// Reverse mappings, codepage -> Unicode
+static std::map<uint16_t, code_page_mapping_reverse_t> mappings_reverse_by_codepage = {};
 
 // ***************************************************************************
 // Grapheme type implementation
 // ***************************************************************************
 
 static bool is_combining_mark(const uint32_t code_point)
 {
-    static const std::vector<std::pair<uint16_t, uint16_t>> ranges = {
-        { 0x0300, 0x036f }, // Combining Diacritical Marks
-        { 0x1ab0, 0x1aff }, // Combining Diacritical Marks Extended
-        { 0x1dc0, 0x1dff }, // Combining Diacritical Marks Supplement
-        { 0x20d0, 0x20ff }, // Combining Diacritical Marks for Symbols
-        { 0xfe20, 0xfe2f }, // Combining Half Marks
-    };
-
-    for (const auto &range : ranges)
-        if (code_point >= range.first && code_point <= range.second)
-            return true;
-
-    return false;
+	static const std::vector<std::pair<uint16_t, uint16_t>> ranges = {
+	        {0x0300, 0x036f}, // Combining Diacritical Marks
+	        {0x064b, 0x065f}, // Arabic Combining Marks
+	        {0x1ab0, 0x1aff}, // Combining Diacritical Marks Extended
+	        {0x1dc0, 0x1dff}, // Combining Diacritical Marks Supplement
+	        {0x20d0, 0x20ff}, // Combining Diacritical Marks for Symbols
+	        {0xfe20, 0xfe2f}, // Combining Half Marks
+	};
+
+	for (const auto& range : ranges) {
+		if (code_point >= range.first && code_point <= range.second) {
+			return true;
+		}
```

Comment:
Just a comment; this is a nice use of `std::any_of`

*Is the code_point in any_of the ranges?*

```c++
auto in_range = [code_point](const auto& range) {
	return code_point >= range.first && code_point <= range.second;
};
return std::any_of(std::begin(ranges), std::end(ranges), in_range);
```

Even though I still find lambda's a bit hard on the eyes, the `any_of` algorithm is a more "code by intent" (and the compiler is then free to implement the search in parallel, etc).

Reply:
Ok, changed.